book/crud.rst

```diff
@@ -0,0 +1,477 @@
+    time.
+
+``filters``
+    Configures the filters displayed above the datagrid.
+
+``actions``
+    Here, you define which actions can be performed by the user. For the links, you refer to the
+    ones defined with the ``properties`` key before. Labels will also be translated. If the
+    configured ``type`` is ``navigate``, clicking the icon will be the same as clicking on an HTML
+    link while using ``delete`` performs an HTTP DELETE request in the background.
+
+.. _book-crud-create-update:
+
+Creating and Updating Entities
+------------------------------
+
+To be able to create new vehicles and update existing ones, you first have to create a form type
+and register it as a service:
+
+.. code-block:: php
+    :linenos:
+
+    // src/InventoryBundle/Form/Type/VehicleType.php
+    namespace InventoryBundle\Form\Type;
+
+    use Symfony\Component\Form\AbstractType;
+    use Symfony\Component\Form\FormBuilderInterface;
+    use Symfony\Component\OptionsResolver\OptionsResolverInterface;
+
+    class VehicleType extends AbstractType
+    {
+        public function buildForm(FormBuilderInterface $builder, array $options)
+        {
+            $builder
+                ->add('model')
+                ->add('seats')
+                ->add('boughtAt')
+                ->add('leasedUntil')
+            ;
+        }
+
+        public function setDefaultOptions(OptionsResolverInterface $resolver)
+        {
+            $resolver->setDefaults(array(
+                'data_class' => 'InventoryBundle\Entity\Vehicle',
+            ));
+        }
+
+        public function getName()
+        {
+            return 'inventory.vehicle';
+        }
+    }
+
+.. code-block:: yaml
+    :linenos:
+
+    # src/InventoryBundle/Resources/config/form.yml
+    services:
+        inventory.form.type.vehicle:
+            class: InventoryBundle\Form\Type\VehicleType
+            tags:
+                - { name: form.type, alias: inventory.vehicle }
+
+Then, you will need to create the needed controller actions. You can simplify the actions if you
+let the ``createAction`` submit the form to the ``updateAction``. Therefore, you have to make sure
+that the ``updateAction`` accepts requests without an existing ``Vehicle`` object being passed:
+
+.. code-block:: php
+    :linenos:
+
+    // src/InventoryBundle/Controller/VehicleController.php
+    namespace InventoryBundle\Controller;
+
+    // ...
+    use InventoryBundle\Entity\Vehicle;
+
+    class VehicleController extends Controller
+    {
+        /**
+         * @Route("/create", name="inventory.vehicle_create")
+         * @Template("InventoryBundle:Vehicle:update.html.twig")
+         * @Acl(
+         *     id="inventory.vehicle_create",
+         *     type="entity",
+         *     class="InventoryBundle:Vehicle",
+         *     permission="CREATE"
+         * )
+         */
+        public function createAction()
+        {
+            return array(
```

Comment:
If we are accepting an empty "Vehicle" object in "updateAction" than we don't really need the "createAction". We'll also need to relink to "inventory.vehicle_update" in "InventoryBundle:Vehicle:index.html.twig"
Another possibility is to put the common logic in a separate method, e.g.:
```php
public function createAction()
{
    return $this->update(new Vehicle());
}

    
public function updateAction(Vehicle $vehicle)
{
    return $this->update($vehicle);
}

protected function update(Vehicle $vehicle)
{
    $form = $this->get('orocrm_contact_us.contact_vehicle.form');

    return array(
        'entity' => $vehicle,
        'form' => $form->createView(),
    );
}
```